src/adapter/src/client.rs

```diff
@@ -235,9 +247,59 @@ pub struct SessionClient {
     cancel_tx: Arc<watch::Sender<Canceled>>,
     cancel_rx: watch::Receiver<Canceled>,
     timeouts: Timeout,
+    segment_client: Option<mz_segment::Client>,
+    environment_id: EnvironmentId,
 }
 
 impl SessionClient {
+    /// Parses a SQL expression, reporting failures as a telemetry event if
+    /// possible.
+    pub fn parse(
+        &self,
+        sql: &str,
+    ) -> Result<Result<Vec<Statement<Raw>>, ParserStatementError>, String> {
+        match mz_sql::parse::parse_with_limit(sql) {
+            Ok(Err(e)) => {
+                self.track_statement_parse_failure(&e);
+                Ok(Err(e))
+            }
+            r => r,
+        }
+    }
+
+    fn track_statement_parse_failure(&self, parse_error: &ParserStatementError) {
+        let session = self.session.as_ref().expect("session invariant violated");
+        let Some(user_id) = session.user().external_metadata.as_ref().map(|m| m.user_id) else {
+            return;
+        };
+        let Some(segment_client) = &self.segment_client else {
+            return;
+        };
+        let Some(statement_kind) = parse_error.statement else {
+            return;
+        };
```

Comment:
It's probably not a big deal, but there are multiple cases in the parser where we know that the statement is a DDL, we just don't know what specific DDL statement kind it is. Those failures are not captured and it might be useful to capture them.